rules/opentitan_test.bzl

```diff
@@ -452,3 +454,138 @@ def opentitan_functest(
+file test_rom.elf
+break rom_test_main
+continue
+delete breakpoints
+
+echo :::: Load the SRAM program onto the device and check integrity.\\n
+file sram_program.elf
+load sram_program.elf
+compare-sections
+
+echo :::: Call sram_main().\\n
+print sram_main()
+
+echo :::: Done.\\n
+    """
+
+    test_script_file = ctx.actions.declare_file("{}.sh".format(ctx.label.name))
+    gdb_script_file = ctx.actions.declare_file("{}.gdb".format(ctx.label.name))
+
+    ctx.actions.write(output = test_script_file, content = test_script)
+    ctx.actions.write(output = gdb_script_file, content = gdb_script)
+
+    return [
+        DefaultInfo(
+            executable = test_script_file,
+            runfiles =
+                ctx.runfiles(
+                    # Relative paths provided by `File.path` seem to not work
+                    # for generated files. Symlinking the runtime files wherever
+                    # Bazel will run `test_script_file` sidesteps the issue.
+                    symlinks = {
+                        "gdb_script": gdb_script_file,
+                        "opentitantool": ctx.file.opentitantool,
+                        "sram_program.elf": ctx.file.sram_program_elf,
+                        "test_rom.bit": ctx.file.test_rom_bitstream,
+                        "test_rom.elf": ctx.file.test_rom_elf,
+                    },
+                    files = [
+                        gdb_script_file,
+                        ctx.file.sram_program_elf,
+                        ctx.file.test_rom_bitstream,
+                        ctx.file.test_rom_elf,
+                        ctx.file.opentitantool,
+                        ctx.file.openocd_earlgrey_config,
+                    ],
+                ),
+        ),
+    ]
+
+# Orchestrate opentitantool, OpenOCD, and GDB to load the given program into
+# SRAM and execute it in-place. This rule assumes that a CW310 FPGA and an
+# ARM-USB-TINY-H JTAG debugger are attached to the host.
+opentitan_sram_fpga_cw310_test = rv_rule(
+    implementation = _opentitan_sram_fpga_cw310_test,
+    attrs = {
+        "sram_program_elf": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+        ),
+        "exit_success_pattern": attr.string(
+            default = "sram_program\\.c:47\\] PC: 0x1000011c, SRAM: \\[0x10000000, 0x10020000\\)",
+        ),
+        "test_rom_bitstream": attr.label(
+            default = "//hw/bitstream:test_rom",
+            allow_single_file = True,
+        ),
+        "test_rom_elf": attr.label(
+            default = "//sw/device/lib/testing/test_rom:test_rom_fpga_cw310.elf",
+            allow_single_file = True,
+        ),
+        "opentitantool": attr.label(default = "//sw/host/opentitantool", allow_single_file = True, cfg = "exec"),
+        "openocd_earlgrey_config": attr.label(
+            default = "//util/openocd/target:lowrisc-earlgrey.cfg",
+            allow_single_file = True,
+        ),
```

Comment:
nit: do you think we will need to customize these?

Reply:
I doubt that `opentitantool` and `openocd_earlgrey_config` will be customized. WDYT about renaming them to `_opentitantool` and `_openocd_earlgrey_config`?

Reply:
SG. Is there a reason for not moving these to the rule?

Reply:
Is there another way to access targets inside the rule implementation? I'm still learning Bazel :)